File: packages/common/services/beautify_service.py
"""
Beautify Service
Main orchestrator for PPTX beautification workflow
KISS: Simple service functions following existing patterns
"""
import logging
import uuid
import secrets
from typing import Any

from sqlalchemy.orm import Session

from packages.common.models.beautify import BeautifySession
from packages.common.models.user import User
from packages.common.core.exceptions import NotFoundError, AuthorizationError
from packages.common.services.pptx_parser import parse_pptx
from packages.common.services.slide_classifier import (
    classify_slide,
    get_suggested_layout,
    get_suggested_alignment,
)
from packages.common.services.mess_analyzer import analyze_mess, calculate_overall_mess_score
from packages.common.services.beautify_transform import transform_slides

logger = logging.getLogger(__name__)


def create_session(
    db: Session,
    user: User,
    file_name: str,
    file_size: int,
    file_key: str | None = None,
) -> BeautifySession:
    """
    Create a new beautify session.
    """
    session = BeautifySession(
        owner_id=user.id,
        file_name=file_name,
        file_size=file_size,
        original_file_key=file_key,
        status="uploading",
        progress=0,
    )
    db.add(session)
    db.commit()
    db.refresh(session)
    return session


def get_session(
    db: Session,
    session_id: uuid.UUID,
    user: User | None = None,
) -> BeautifySession:
    """
    Get session with optional authorization check.
    """
    session = db.query(BeautifySession).filter(
        BeautifySession.id == session_id
    ).first()

    if not session:
        raise NotFoundError(
            message="Beautify session not found",
            resource_type="BeautifySession",
            resource_id=str(session_id),
        )

    if user and session.owner_id != user.id:
        raise AuthorizationError("You do not have access to this session")

    return session


def update_session_status(
    db: Session,
    session_id: uuid.UUID,
    status: str,
    progress: int = 0,
    error_message: str | None = None,
) -> None:
    """
    Update session status and progress.
    """
    session = db.query(BeautifySession).filter(
        BeautifySession.id == session_id
    ).first()

    if session:
        session.status = status
        session.progress = progress
        if error_message:
            session.error_message = error_message
        db.commit()


def process_pptx_file(
    db: Session,
    session_id: uuid.UUID,
    file_path: str,
) -> list[dict[str, Any]]:
    """
    Process uploaded PPTX file:
    1. Parse PPTX to extract raw slides
    2. Classify each slide
    3. Analyze mess scores
    4. Build SlideIR list

    Returns list of SlideIR dicts.
    """
    # Update status: parsing
    update_session_status(db, session_id, "parsing", 10)

    try:
        # Parse PPTX
        raw_slides = parse_pptx(file_path)
        total_slides = len(raw_slides)

        if total_slides == 0:
            raise ValueError("No slides found in PPTX file")

        update_session_status(db, session_id, "analyzing", 30)

        # Process each slide
        slides_ir = []
        for idx, raw_slide in enumerate(raw_slides):
            # Classify slide
            slide_type, confidence = classify_slide(raw_slide)

            # Analyze mess
            mess_score, mess_issues = analyze_mess(raw_slide)

            # Get suggestions
            suggested_layout = get_suggested_layout(slide_type)
            suggested_alignment = get_suggested_alignment(slide_type)

            # Extract title and content
            title = extract_title(raw_slide)
            content = extract_content(raw_slide)

            # Get first image URL if exists
            images = raw_slide.get("images", [])
            image_url = images[0].get("data_url") if images else None

            # Build SlideIR
            slide_ir = {
                "id": str(uuid.uuid4()),
                "position": idx,
                "type": slide_type,
                "typeConfidence": confidence,
                "title": title,
                "content": content,
                "notes": raw_slide.get("notes", ""),
                "imageUrl": image_url,
                "messScore": mess_score,
                "messIssues": mess_issues,
                "suggestedLayout": suggested_layout,
                "suggestedAlignment": suggested_alignment,
                "original": {
                    "texts": raw_slide.get("texts", []),
                    "images": [
                        {"width": img.get("width"), "height": img.get("height")}
                        for img in images
                    ],  # Don't store full base64 in DB
                    "notes": raw_slide.get("notes"),
                    "backgroundColor": raw_slide.get("background_color"),
                },
            }
            slides_ir.append(slide_ir)

            # Update progress
            progress = 30 + int((idx + 1) / total_slides * 60)
            update_session_status(db, session_id, "analyzing", progress)

        # Calculate overall mess score
        overall_mess = calculate_overall_mess_score(raw_slides)

        # Update session with results
        session = get_session(db, session_id)
        session.slides_data = slides_ir
        session.overall_mess_score = overall_mess
        session.status = "ready"
        session.progress = 100
        db.commit()

        return slides_ir

    except Exception as e:
        logger.error(f"Error processing PPTX: {e}")
        update_session_status(db, session_id, "error", 0, str(e))
        raise


def transform_session(
    db: Session,
    session_id: uuid.UUID,
    user: User,
    theme_id: str,
    intensity: str,
) -> list[dict[str, Any]]:
    """
    Transform session slides with selected style and intensity.
    """
    session = get_session(db, session_id, user)

    if session.status not in ["ready", "done"]:
        raise ValueError(f"Session not ready for transform. Status: {session.status}")

    slides_ir = session.slides_data or []
    if not slides_ir:
        raise ValueError("No slides to transform")

    # Update status
    session.status = "transforming"
    db.commit()

    try:
        # Transform slides
        transformed = transform_slides(slides_ir, theme_id, intensity)

        # Save results
        session.theme_id = theme_id
        session.intensity = intensity
        session.transformed_slides = transformed
        session.status = "done"
        db.commit()

        return transformed

    except Exception as e:
        logger.error(f"Error transforming slides: {e}")
        session.status = "error"
        session.error_message = str(e)
        db.commit()
        raise


def create_share_link(
    db: Session,
    session_id: uuid.UUID,
    user: User,
) -> tuple[str, str]:
    """
    Generate unique share ID and URL.
    """
    session = get_session(db, session_id, user)

    if session.status != "done":
        raise ValueError("Session must be completed before sharing")

    # Generate share ID if not exists
    if not session.share_id:
        session.share_id = secrets.token_urlsafe(12)
        session.is_public = True
        db.commit()

    # Generate share URL (frontend will handle the actual URL)
    share_url = f"/share/beautify/{session.share_id}"

    return session.share_id, share_url


def get_share_data(
    db: Session,
    share_id: str,
) -> dict[str, Any]:
    """
    Get public share view data.
    """
    session = db.query(BeautifySession).filter(
        BeautifySession.share_id == share_id,
        BeautifySession.is_public == True,
    ).first()

    if not session:
        raise NotFoundError(
            message="Share link not found or expired",
            resource_type="BeautifySession",
            resource_id=share_id,
        )

    return {
        "file_name": session.file_name,
        "before_slides": session.slides_data or [],
        "after_slides": session.transformed_slides or [],
        "theme_id": session.theme_id or "executive",
        "created_at": session.created_at,
    }


def extract_title(raw_slide: dict[str, Any]) -> str:
    """Extract title from raw slide data."""
    texts = raw_slide.get("texts", [])

    # Find text marked as title
    for text in texts:
        if text.get("is_title"):
            return text.get("value", "").strip()

    # Fall back to largest font
    if texts:
        largest = max(texts, key=lambda t: t.get("font_size", 0))
        return largest.get("value", "").strip()[:200]  # Limit length

    return "Untitled"


def extract_content(raw_slide: dict[str, Any]) -> list[str]:
    """Extract content bullets from raw slide data."""
    texts = raw_slide.get("texts", [])
    content = []

    for text in texts:
        if text.get("is_title"):
            continue

        value = text.get("value", "").strip()
        if not value:
            continue

        # Split by newlines if present
        if "\n" in value:
            lines = [line.strip() for line in value.split("\n") if line.strip()]
            content.extend(lines)
        else:
            content.append(value)

    return content[:20]  # Limit to 20 items


def delete_session(
    db: Session,
    session_id: uuid.UUID,
    user: User,
) -> None:
    """Delete a beautify session."""
    session = get_session(db, session_id, user)
    db.delete(session)
    db.commit()

File: services/api/beautifyService.ts
/**
 * Beautify Service
 *
 * API client for the "Make My Ugly Deck Beautiful" feature.
 * SRP: Single service for all beautify-related API calls.
 */

import { api } from './apiClient';
import { API_BASE_URL } from '@/config';
import type { Slide } from '@/types';
import type {
  BeautifySession,
  SlideIR,
  TransformIntensity,
  UploadResponse,
  TransformResponse,
  ShareResponse,
  ShareViewData,
} from '@/types/beautify';

// =============================================================================
// BACKEND TYPES (snake_case)
// =============================================================================

interface BackendSlideIR {
  id: string;
  position: number;
  type: string;
  type_confidence: number;
  title: string;
  content: string[];
  notes: string;
  image_url?: string;
  mess_score: number;
  mess_issues: string[];
  suggested_layout: string;
  suggested_alignment: string;
  original: {
    texts: Array<{
      value: string;
      font_size?: number;
      font_family?: string;
      is_bold?: boolean;
      is_title?: boolean;
    }>;
    images: Array<{
      url: string;
      width?: number;
      height?: number;
    }>;
    notes?: string;
    background_color?: string;
  };
}

interface BackendSession {
  id: string;
  file_name: string;
  file_size: number;
  created_at: number;
  slides: BackendSlideIR[];
  intensity: string;
  theme_id: string;
  overall_mess_score: number;
  status: string;
  progress: number;
  error_message?: string;
  share_id?: string;
}

// =============================================================================
// CONVERSION HELPERS
// =============================================================================

function toSlideIR(backend: BackendSlideIR): SlideIR {
  return {
    id: backend.id,
    position: backend.position,
    type: backend.type as SlideIR['type'],
    typeConfidence: backend.type_confidence,
    title: backend.title,
    content: backend.content,
    notes: backend.notes,
    imageUrl: backend.image_url,
    messScore: backend.mess_score,
    messIssues: backend.mess_issues,
    suggestedLayout: backend.suggested_layout as SlideIR['suggestedLayout'],
    suggestedAlignment: backend.suggested_alignment as SlideIR['suggestedAlignment'],
    original: {
      texts: backend.original.texts.map((t) => ({
        value: t.value,
        fontSize: t.font_size,
        fontFamily: t.font_family,
        isBold: t.is_bold,
        isTitle: t.is_title,
      })),
      images: backend.original.images,
      notes: backend.original.notes,
      backgroundColor: backend.original.background_color,
    },
  };
}

function toBeautifySession(backend: BackendSession): BeautifySession {
  return {
    id: backend.id,
    fileName: backend.file_name,
    fileSize: backend.file_size,
    createdAt: backend.created_at,
    slides: backend.slides.map(toSlideIR),
    intensity: backend.intensity as TransformIntensity,
    themeId: backend.theme_id,
    overallMessScore: backend.overall_mess_score,
    status: backend.status as BeautifySession['status'],
    progress: backend.progress,
    errorMessage: backend.error_message,
    shareId: backend.share_id,
  };
}

// =============================================================================
// API FUNCTIONS
// =============================================================================

/**
 * Upload a PPTX file for beautification.
 * Returns a session ID for tracking progress.
 */
export async function uploadPptx(file: File): Promise<UploadResponse> {
  const formData = new FormData();
  formData.append('file', file);

  // Use fetch directly for FormData (api client JSON-ifies body)
  const response = await fetch(
    `${API_BASE_URL}/api/v1/beautify/upload`,
    {
      method: 'POST',
      body: formData,
      credentials: 'include',
    }
  );

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Upload failed' }));
    throw new Error(error.detail || 'Upload failed');
  }

  const data = await response.json();
  return { sessionId: data.session_id };
}

/**
 * Get the current state of a beautify session.
 * Poll this while status is 'parsing' or 'analyzing'.
 */
export async function getSession(sessionId: string): Promise<BeautifySession> {
  const backend = await api.get<BackendSession>(`/api/v1/beautify/${sessionId}`);
  return toBeautifySession(backend);
}

/**
 * Transform the uploaded deck with selected style and intensity.
 * Returns transformed slides using the existing Slide type.
 */
export async function transformDeck(
  sessionId: string,
  themeId: string,
  intensity: TransformIntensity
): Promise<TransformResponse> {
  const response = await api.post<{ slides: Slide[] }>(
    `/api/v1/beautify/${sessionId}/transform`,
    { theme_id: themeId, intensity }
  );
  return { slides: response.slides };
}

/**
 * Create a shareable before/after link.
 */
export async function createShareLink(sessionId: string): Promise<ShareResponse> {
  const response = await api.post<{ share_id: string; share_url: string }>(
    `/api/v1/beautify/${sessionId}/share`
  );
  return {
    shareId: response.share_id,
    shareUrl: response.share_url,
  };
}

/**
 * Get public share data for a before/after view.
 * This endpoint doesn't require authentication.
 */
export async function getShareData(shareId: string): Promise<ShareViewData> {
  const response = await api.get<{
    file_name: string;
    before_slides: BackendSlideIR[];
    after_slides: Slide[];
    theme_id: string;
    created_at: number;
  }>(`/api/v1/share/beautify/${shareId}`, { skipAuth: true });

  return {
    fileName: response.file_name,
    beforeSlides: response.before_slides.map(toSlideIR),
    afterSlides: response.after_slides,
    themeId: response.theme_id,
    createdAt: response.created_at,
  };
}

// =============================================================================
// CONVENIENCE EXPORT
// =============================================================================

export const beautifyService = {
  upload: uploadPptx,
  getSession,
  transform: transformDeck,
  share: createShareLink,
  getShareData,
};
